UWP native code docs (#270)
microsoft/react-native-windows-samples: docs/native-code.md

---
id: native-code
title: Working with native code on Windows
---

## What is a React Native for Windows app?

> A RNW app is a [Universal Windows Platform app](https://docs.microsoft.com/en-us/windows/uwp/get-started/universal-application-platform-guide) (aka UWP or Windows 10 app).

Reviewer: ```suggestion
A RNW app is a [Universal Windows Platform app](https://docs.microsoft.com/windows/uwp/get-started/universal-application-platform-guide) (aka UWP or Windows 10 app).

```

Throughout, please remove the language tokens from the URLs so that folks visiting from other locales get the right version of the sites.

The Universal Windows Platform allows you to access a set of common functionality on all Windows devices via the [Windows Runtime](https://docs.microsoft.com/en-us/windows/uwp/winrt-components/) (WinRT). WinRT APIs can be accessed from C++ (via [C++/WinRT](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/)), or via .NET C#. 

#### WinRT support in .NET

> The current publicly supported version of .NET has built-in support for WinRT ([.NET UWP](https://docs.microsoft.com/en-us/windows/uwp/winrt-components/creating-windows-runtime-components-in-csharp-and-visual-basic)) whereas future versions of .NET Core starting with .NET Core 5 will only be able to access WinRT APIs via [C#/WinRT](https://docs.microsoft.com/en-us/windows/uwp/csharp-winrt/) in a way similar to what the C++/WinRT projection enables today. 

Reviewer: ```suggestion
The current publicly supported version of .NET has built-in support for WinRT ([.NET UWP](https://docs.microsoft.com/en-us/windows/uwp/winrt-components/creating-windows-runtime-components-in-csharp-and-visual-basic)). 

Future versions of .NET Core starting with .NET Core 5 will only be able to access WinRT APIs via [C#/WinRT](https://docs.microsoft.com/en-us/windows/uwp/csharp-winrt/) in a way similar to what the C++/WinRT projection enables today.
```

Should we discuss future versions of .NET Core in this doc? I don't understand what the developer action is as a result of this sentence. Is it that their native modules must be written in CsWinRT?

Author: .NET Core 5 is about to go live (they already published RC 2). Once they go live, UWP apps written in managed languages should likely be written in .NET Core 5 as .NET UWP will be legacy/deprecated.

Reviewer: I still don't understand what the impact is for RNW developers. Will they need to do something special when .NET Core 5 goes live? Or will that be handled by the RNW platform? Will they need to update their native modules?

Author: Hmm ok. There still is no official guidance re: .net core 5 and UWP (e.g. https://github.com/microsoft/ProjectReunion/issues/105), it seems like .net core 5 might not support UWP apps "as we know them today" out of the box, so maybe we should have a discussion about what this means separate from this PR.

### API surface
The set of APIs these app can access are a subset of all Windows APIs (i.e. those accessible via WinRT). See:
- [Win32 and COM APIs for UWP apps](https://docs.microsoft.com/en-us/uwp/win32-and-com/win32-and-com-for-uwp-apps)
- [CRT functions not supported in Universal Windows Platform apps](https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps)
- [Alternatives to Windows APIs in Universal Windows Platform (UWP) apps](https://docs.microsoft.com/en-us/uwp/win32-and-com/alternatives-to-windows-apis-uwp)

### Isolation
The app runs inside of an app container - a type of sandbox. This provides apps with a secure way to install, access system resources like the filesystem, and lets the system manage their lifetime (e.g. suspending the app when it isn't on the foreground). This means that by default an RNW app cannot access arbitrary filesystem locations, start arbitrary processes, etc. UWP apps that need to access these kinds of capabilities may be able to do so via [App capability declarations](https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations).

### Packaging

> They are signed and packaged. [Packaging](https://docs.microsoft.com/en-us/windows/uwp/packaging/) is a mechanism through which an app and its dependencies acquire an identity, which is used to determine whether API calls that require system capabilities (e.g. filesystem access) should succeed or not.

Reviewer: ```suggestion
React Native Windows apps are signed and packaged. [Packaging](https://docs.microsoft.com/en-us/windows/uwp/packaging/) is a mechanism through which an app and its dependencies acquire an identity, which is used to determine whether API calls that require system capabilities (e.g. filesystem access) should succeed or not.

```

### Distribution

> They can be distributed, installed and updated in the following ways:

Reviewer: ```suggestion
React Native Windows apps can be distributed, installed and updated in the following ways:

```

  - via [the Microsoft Store](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-distribute).
  - via [your private Store](https://docs.microsoft.com/en-us/microsoft-store/distribute-apps-to-your-employees-microsoft-store-for-business) if you are a business or educational organization. See also [Distribute LOB apps to enterprises](https://docs.microsoft.com/en-us/windows/uwp/publish/distribute-lob-apps-to-enterprises).
  - using [App Installer](https://docs.microsoft.com/en-us/windows/msix/app-installer/installing-windows10-apps-web).

>   - It's worth noting that you cannot just "_copy an exe_" as the app package contains more than just the main executable, including an app manifest, assets, dependent framework libraries, etc.

Reviewer: This bullet and later bullets don't follow the same pattern as the first three; should this be a separate list? It seems like these bullets are more FYI whereas the first bullets are about how the apps can be distributed.

- UWP apps submitted to the store must pass [Windows App Certification Kit](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/windows-app-certification-kit) (WACK) validation.
- UWP apps written in C# or other managed languages submitted to the store must be built using the [.NET Native toolchain](https://docs.microsoft.com/en-us/dotnet/framework/net-native/). This is the default when building C# apps in Release mode, **but not in Debug**, so apps built in Debug will be rejected by the Store.

### Use of non-WinRT libraries
Any libraries you use should be built as WinRT components. In other words, you cannot easily link libraries built for win32 desktop apps without additional work.

>   - C++/CX is a dialect of C++ that allows writing UWP apps, however this is **not supported** for writing a RNW app. The article [How to use existing C++ code in a Universal Windows Platform app](https://docs.microsoft.com/en-us/cpp/porting/how-to-use-existing-cpp-code-in-a-universal-windows-platform-app) talks about how to consume non-WinRT libraries in a WinRT context using C++/CX, but most of the content should be applicable to using C++/WinRT which is the supported way to write RNW apps.

Reviewer: I'm confused about this bullet as it relates to the native modules docs, which I thought could be written in "pure" C++. 

Meta: I'm not sure what the relationship between these docs and the native modules docs all-up, especially this section.

Author: Native modules allow you to access the underlying platform, which for UWP you would do via cppwinrt. Yes, technically you can write a pure C++ module without calling any winrt components (e.g. the sample add x+y native module), but that's not what people will usually do.

The point of this paragraph is about "bring some library from nuget". We've seen people ask this on the repo, "can I use this win32 library in the context of a native module?", or they try it and it doesn't work. The info here is to warn them about it and provide some paths forward.

  - Libraries built for .NET desktop framework cannot be directly accessed by UWP. You can create a .NET Standard library that calls into the .NET framework one, and call from the UWP app into the .NET Standard middleware.

### Local testing and inner loop
For internal development, you can deploy your app for test purposes by side-loading and deploying via [loose-file registration](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/loose-file-registration). When building in **Debug** mode (which is the default), `npx react-native run-windows` performs loose-file registration of your app in order to install it locally. When running `npx react-native run-windows` with the `--release` switch, the CLI will install the real package onto your local machine. This requires the app to be signed and for the certificate it uses for signing to be trusted by the machine the app is going to be installed on. See [Create a certificate for package signing](https://docs.microsoft.com/en-us/windows/msix/package/create-certificate-package-signing) and [Intro to certificates](https://docs.microsoft.com/en-us/windows/uwp/security/certificates).

> ## C# vs. C++ for Native Modules

Reviewer: It looks like this was cut-pasted from https://microsoft.github.io/react-native-windows/docs/native-modules#c-vs-c-for-native-modules  but some content is lost (esp. around the CLR, the specific performance implications, etc.) - where did that content go?

Author: I deleted the content about loading the CLR since that is not true for UWP apps as they are cross compiled to native.

Reviewer: That's fine - we should still articulate how the choice of language impacts performance or remove this line.

### Packaging

React Native Windows apps are signed and packaged. [Packaging](https://docs.microsoft.com/windows/uwp/packaging/) is a mechanism through which an app and its dependencies acquire an identity, which is used to determine whether API calls that require system capabilities (e.g. filesystem access) should succeed or not.

### Distribution
React Native Windows apps can be distributed, installed and updated in the following ways:

  - via [the Microsoft Store](https://docs.microsoft.com/windows/apps/desktop/modernize/desktop-to-uwp-distribute).
  - via [your private Store](https://docs.microsoft.com/microsoft-store/distribute-apps-to-your-employees-microsoft-store-for-business) if you are a business or educational organization. See also [Distribute LOB apps to enterprises](https://docs.microsoft.com/windows/uwp/publish/distribute-lob-apps-to-enterprises).
  - using [App Installer](https://docs.microsoft.com/windows/msix/app-installer/installing-windows10-apps-web).

It's worth noting that you cannot just "_copy an exe_" as the app package contains more than just the main executable, including an app manifest, assets, dependent framework libraries, etc.

> In addition, the Store submission process has this requirements:

Reviewer: ```suggestion
In addition, the Store submission process has these requirements:

```

C# development brings engineering efficiencies to writing a module or an app. We strive to maintain parity in developer experiences between C# and C++. If your app or module already uses C#, you should feel empowered to continue to use C#. That said, modules that Microsoft contributes to will be written in C++ to ensure the highest level of performance and interoperability.

<div class="warning">
  <h3>Important</h3>
  <h4>Mixing C# and C++</h4>
</div>

### App crashes when trying to load the C# component or instantiate one of its types

> **Error 0x80131040 “The located assembly’s manifest definition does not match the assembly reference”**

Reviewer 2: I believe the 0.63+ C++ sample template does some of this by default now.

Author: The property exists, it is not on by default as it would bloat fully native apps unnecessarily.

To work around this problem there are three options:
1. Set your C# component's target Windows version to Windows 10 version 1703 (Build 15063) or lower.
1. Reference the .NET Native nuget packages in your C++ app:
   - Right click on the app's .vcxproj file → **Manage NuGet Packages**.
   - Search for `Microsoft.Net.Native.Compiler`, and install it.
   - Then add the following properties to the .vcxproj file:
    ```xml
    <PropertyGroup>
      <UseDotNetNativeToolchain Condition="'$(Configuration)'=='Release'">true</UseDotNetNativeToolchain>
      <DotNetNativeVersion>2.2.3</DotNetNativeVersion>
    </PropertyGroup>
    ```
1. In your .vcxproj file, set this property in the first `<PropertyGroup>`:
   ```xml
   <ConsumeCSharpModules>true</ConsumeCSharpModules>
   ```

## Troubleshooting and debugging C++ native modules

So you added a new native module or a new method to a module but it isn't working, **now what?!**

If your method isn't being hit in the VS debugger, something is blocking the call due to a mismatch, likely between the expected and actual types that your method takes/returns.

> To debug into what is rejecting the call, set a breakpoint in `CxxNativeModule::invoke` (See [ReactCommon\react-native-patched\ReactCommon\cxxreact\CxxNativeModule.cpp](https://github.com/facebook/react-native/blob/0b8a82a6eeeb3508b80ee137d313f64fe323db06/ReactCommon/cxxreact/CxxNativeModule.cpp#L97)). This breakpoint is bound to be hit a lot (every time a call to a native method is made), so we want to make sure we only break when *our* method of interest is involved.

Reviewer 2: We probably don't want to point folks towards this. react-native-patched is a build time directory we're trying to remove (it will be gone on binary distribution as well), and Facebook has been refactoring some of this.

Author: We can update the docs when the file gets removed.
